kvdb/postgres/readwrite_bucket.go

```diff
@@ -427,3 +427,31 @@ func (b *readWriteBucket) Sequence() uint64 {
 
 	return uint64(seq)
 }
+
+func (b *readWriteBucket) Prefetch(paths ...[]string) {}
+
+func (b *readWriteBucket) ForEachFast(cb func(k, v []byte) error) error {
```

Comment:
How about naming this `ForAll`? Since it'd be a valid question to ask why `ForEach` is not as fast as `ForEachFast`.

Reply:
Hm naming is difficult. `ForEach` and `ForAll` also sounds like the same thing to me. I guess the functional difference is that the faster one doesn't allow further queries in the callback.

Reply:
Aha, I missed the not allowing more queries in the callback part. What I've expected instead is that it'll simply batch fetch more items at once under the hood.

Reply:
Also weighing in on the naming discussion :) How about `ForEachUnordered` (the main promise of the cursor based `ForEach` is the sort by key and seek functionality, right?) or `ForEachNative` (implying that the underlying implementation might decide on the best approach for executing it, could also mention no specific ordering is guaranteed).

Reply:
This is what I wanted to do originally with the cursor. In that case, no new method is needed either. But I remember there was a problem with that, only what was it?

I don't think it was writing to the bucket inside the loop, because that is already not supported with bbolt - or is it? Don't we have these iterate and delete loops?

Reply:
The batching with the cursor may get a bit ugly because how many items to pre-load? But just for `ForEach` in a read-only tx, it should be possible I think. Will take a look, nice suggestion :+1:

Reply:
> The batching with the cursor may get a bit ugly because how many items to pre-load? And then there is also the possibility to go forward and backward. But just for ForEach in a read-only tx, it should be possible I think. Will take a look, nice suggestion 

I've implemented this once and it's indeed ugly. For specific use cases where we fetch a whole bucket as fast as possible, I'd say we could overcome the bolt emulation altogether and just do something like `tx.GetAllPaginated(path, limit, ordered)` where the path is similarly defined as in the `Prefetch`

Reply:
This would be compatible with both implementations and would avoid caching things in the STM as well.

Reply:
Interesting. So this would also allow nested keys to be retrieved in a single query. And besides `limit` you'd also want to specify some kind of `fromKey` right?

Reply:
Or alternatively we can use `start` (index). That would translate to `OFFSET .. LIMIT ..` in sql.

Reply:
I think it is still the case with `GetAllPaginated` that the callback isn't allowed to do other queries.

Reply:
Or maybe your idea is to not have a callback with `GetAllPaginated` and return `[]KV` ?

Reply:
> Interesting. So this would also allow nested keys to be retrieved in a single query. And besides `limit` you'd also want to specify some kind of `fromKey` right?

Yes.

> Or alternatively we can use start (index). That would translate to OFFSET .. LIMIT .. in sql.

I think `fromKey` is more "portable".

> I think it is still the case with GetAllPaginated that the callback isn't allowed to do other queries, so maybe this is just scope creep beyond a single ForAll?

Maybe we should go even further and add it to the `kvdb.DB` interface as an extension. That'd indicate that it's a pure optimization thing and is not needed to be compatible with the rest of the `kvdb` interfaces.

> Or maybe your idea is to not have a callback with GetAllPaginated and return []KV ?

Yeah, sgtm!

Reply:
I get the idea, but after sleeping on it, I think we should do the minimum that is necessary at this point. Just fix the immediate problem with postgres startup loading.

If another critical optimization comes up, we can look at what is needed for that. I don't expect that updating the `ForAllNative` that we use now to something slightly different in the future will be a problem.